ricecooker/classes/nodes.py

```diff
@@ -798,3 +798,84 @@ def truncate_fields(self):
+            thumbnail (str): local path or url to thumbnail image (optional)
+            extra_fields (dict): any additional data needed for node (optional)
+            domain_ns (str): who is providing the content (e.g. learningequality.org) (optional)
+    """
+    kind = content_kinds.SLIDESHOW
+
+    def __init__(self, source_id, title, license, slideshow_data=None, **kwargs):
+        if slideshow_data:
+            extra_fields = {'slideshow_data': slideshow_data}
+        else:
+            extra_fields = {'slideshow_data':[]}
+        # THe Node base class' __init__ method has:
+        #       for f in files or []:
+        #           self.add_file(f)
+        super(SlideshowNode, self).__init__(source_id, title, license, extra_fields=extra_fields, **kwargs)
+
+    def add_file(self, file_to_add):
+        """
+        Add a the SlideImageFile to the node's files and append slideshow metadata
+        to extra_fields['slideshow_data'] (list).
+        Args: file (SlideshowNode or ThumbnailFile): file model to add to node
+        Returns: None
+        """
+        from .files import File, SlideImageFile
+        assert isinstance(file_to_add, File), "Files being added must be instances of a subclass of File class"
+        # TODO asserr SlideshowNode OR ThumbnailFile
+
+        if file_to_add not in self.files:
+            filename = file_to_add.get_filename()
+            if filename:
+                checksum, ext = filename.split('.')  # <md5sum(contents)>.[png|jpg|jpeg]
+            else:
+                raise ValueError('filename not available')
+            #
+            # Find the idx of sort_order.next()
+            slideshow_image_files = [f for f in self.files if isinstance(f,SlideImageFile)]
+            idx = len(slideshow_image_files)  # next available index, assuming added in desired order
+            #
+            # Add slideshow data to extra_fields['slideshow_data'] (aka manifest)
+            slideshow_data = self.extra_fields['slideshow_data']
+            slideshow_data.append(
+                {
+                    'caption': file_to_add.caption,
+                    'descriptive_text': file_to_add.descriptive_text,
+                    'sort_order': idx,
+                    'checksum': checksum,
+                    'extension': ext
+                }
+            )
+            self.extra_fields['slideshow_data'] = slideshow_data
+            #
+            # Add node->file link
+            file_to_add.node = self
+            self.files.append(file_to_add)
+
+    def validate(self):
+        print('in SlideshowNode validate')
+        from .files import SlideImageFile
+        try:
+            assert any(self.files), "Assumption Failed: Slideshow does not have any slideshow image files."
+            assert all(filter(lambda f: isinstance(f, SlideImageFile), self.files)), "Assumption Failed: Slideshow files must all be of type SlideImageFile."
+            # TODO: could also be ThumbnailFile    ^
```

Comment:
Validation will fail if you add a `ThumbnailFile` to the node now --- so need to change line 877 to be an OR check

Reply:
Hey @ivanistheone 

I'm curious if I'm misunderstanding the expected functionality of that lambda line at 877. My understanding is that it should parse all of the `self.files` passing each as `f` to the lambda, which then evaluates the `isinstance(...` expression for each file `f`.

However:

```
assert all(filter(lambda f: isinstance(f, SlideImageFile), self.files)), "Assumption Failed: Slideshow files must all be of type SlideImageFile."
```

Is not working functionally the same as:

```
for file in self.files:
                assert isinstance(file, SlideImageFile), "Assumption Failed: Slideshow files must all be of type SlideImageFile."
```

 Because - when I went into the tests and added ThumbnailFiles to the test Slideshows, there was no failure at all - which meant that the first assert that used the filter is not properly asserting against all files.

The second `for x in y` type example, however, fails as it should when it gets to the ThumbnailFile.

I also encountered this exact same issue elsewhere while writing tests and I'll make the note there.

Reply:
I think the confusion is with the `filter` function which is not needed for this check. Instead what we'd want is something more like: 
```
assert all([isinstance(f, SlideImageFile) or isinstance(f, ThumbnailFile) for f in self.files]), \
               "SlideshowNode files must be of type SlideImageFile or ThumbnailFile."
```
but [your version with the for loop](https://github.com/nucleogenesis/ricecooker/blob/slideshow-content-kind/ricecooker/classes/nodes.py#L881-L882) is just as good—no need to try fit everything in on one line...